patches/api/0374-More-Teleport-API.patch

```diff
@@ -0,0 +1,238 @@
++
++    /**
++     * Causes the player to look towards the given location.
++     *
++     * @param location Location to look at
++     * @param playerAnchor What part of player should face the location
++     */
++    default public void lookAt(@NotNull Location location, @NotNull io.papermc.paper.entity.LookAnchor playerAnchor) {
++        lookAt(location.getX(), location.getY(), location.getZ(), playerAnchor);
++    }
++
++    /**
++     * Causes the player to look towards the given entity.
++     *
++     * @param playerAnchor What part of the player should face the entity
++     * @param entity Entity to look at
++     * @param entityAnchor What part of the entity the player should face
++     */
++    public void lookAt(@NotNull io.papermc.paper.entity.LookAnchor playerAnchor, org.bukkit.entity.@NotNull Entity entity, @NotNull io.papermc.paper.entity.LookAnchor entityAnchor);
++    // Paper end - Teleport API
++
+     @NotNull
+     @Override
+     Spigot spigot();
+diff --git a/src/main/java/org/bukkit/event/player/PlayerTeleportEvent.java b/src/main/java/org/bukkit/event/player/PlayerTeleportEvent.java
+index 553d7740489fe729166c8ca8ef8c7834db3663ad..2aeaa892af957fa4461a3c211ba7724ab14798f1 100644
+--- a/src/main/java/org/bukkit/event/player/PlayerTeleportEvent.java
++++ b/src/main/java/org/bukkit/event/player/PlayerTeleportEvent.java
+@@ -23,6 +23,16 @@ public class PlayerTeleportEvent extends PlayerMoveEvent {
+         this.cause = cause;
+     }
+ 
++    // Paper start - Teleport API
++    private boolean dismounted = true;
++    private java.util.Set<io.papermc.paper.entity.RelativeTeleportFlag> teleportFlagSet = java.util.Collections.emptySet();
++
++    public PlayerTeleportEvent(@NotNull final Player player, @NotNull final Location from, @Nullable final Location to, @NotNull final TeleportCause cause, boolean dismounted, @NotNull java.util.Set<io.papermc.paper.entity.RelativeTeleportFlag> teleportFlagSet) {
++        this(player, from, to, cause);
++        this.dismounted = dismounted;
++        this.teleportFlagSet = teleportFlagSet;
++    }
++    // Paper end
+     /**
+      * Gets the cause of this teleportation event
+      *
+@@ -80,6 +90,28 @@ public class PlayerTeleportEvent extends PlayerMoveEvent {
+         UNKNOWN;
+     }
+ 
++    // Paper start - Teleport API
++    /**
++     * Gets if the player will be dismounted in this teleportation.
++     *
++     * @return dismounted or not
++     */
++    public boolean willDismountPlayer() {
++        return dismounted;
++    }
++
++    /**
++     * Returns the relative teleportation flags used in this teleportation.
++     * This determines which axis the player will not lose their velocity in.
++     *
++     * @return relative teleportation flags
++     */
++    @NotNull
++    public java.util.Set<io.papermc.paper.entity.RelativeTeleportFlag> getRelativeTeleportationFlags() {
```

Comment:
Either we make this set immutable or we clearly state that mutation of the set will not affect teleportation.
If I just had a first glance at the event as a developer and I see a set returned like this (e.g. like blockList in explosion events) I'd assumable mutability and said mutations to affect the outcome.

Technically this could be mutable ? Might just be a point to discuss :+1:

Reply:
Yea, I guess making it mutable is fine.

Reply:
I mean, idk how useful mutability is here. A listener mutating this seems pretty unlikely. Just something to choose here.